### ISSUE TITLE
Help with docker-coq/docker-mathcomp rebuilds: looking for a volunteer

### ISSUE THREAD

**Issue by erikmd** [coq/coq#19306]
**TL;DR:** would a person from @coq/core volunteer to help me to trigger docker-base or docker-coq rebuilds?

## Concretely, the problem:

* {docker-coq, docker-mathcomp} images aim to earn CI time by providing pre-compiled opam switches with coq+bignums(+coq-mathcomp-character).
* around several times per months, it happens some base opam packages are modified upstream, impyling coq is "rebuilt downstream" which thereby reduces the usefulness of the prebuilt coq. See the Example log below that shows the standard log pattern in this case.

## The solution:

* last week-end, I significantly improved the [docker-keeper](https://gitlab.com/erikmd/docker-keeper) tool to allow "transitive rebuilds", and [`propagate strategy`](https://github.com/coq-community/docker-coq/blob/cae8cadccb2266eb15682b9c56ab69671bcffed6/images.yml#L11-L41), and HEAD commit message parsing.
* now, it is as easy as running
  ```
  cd docker-coq
  git pull
  git commit --allow-empty -m "docker-keeper: rebuild-all"
  git push
  ```
  **or** (depending on the **layers** that have to be rebuilt)
  ```
  cd docker-base
  git pull
  git commit --allow-empty -m "docker-keeper: rebuild-all"
  git push
  ```
 * see also the [Zulip thread](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/New.20release.20of.20docker-keeper/near/448382199) which provides more technical details, and the discussion with @palmskog that led to this call
* to sum up, I wouldn't want to be the only person that can trigger rebuilds, hence this volunteer call @coq/core  
  (I also opened a call for math-comp/core)
* the next step would be to post a message on Coq Discourse to tell end users about the typical log pattern below, and the possibility to ping me or the volunteer people.

## Example log (in docker-coq-action CI jobs)

https://github.com/coq-community/graph-theory/actions/runs/9684414743/job/26722132412

```
  The following actions will be performed:
    - recompile conf-gmp               4             [upstream or system changes]
    - recompile zarith                 1.13*         [uses conf-gmp]
    - recompile coq-core               8.18.0        [uses zarith]
    - recompile coqide-server          8.18.0        [uses coq-core]
    - recompile coq-stdlib             8.18.0        [uses coq-core]
    - recompile coq                    8.18.0*       [uses coq-core]
    - recompile coq-serapi             8.18.0+0.18.3 [uses coq]
    - recompile coq-elpi               1.19.3        [uses coq]
    - recompile coq-bignums            9.0.0+coq8.18 [uses coq]
    - recompile coq-hierarchy-builder  1.6.0         [uses coq-elpi]
    - recompile coq-mathcomp-ssreflect 2.0.0*        [uses coq]
    - install   coq-mathcomp-finmap    2.1.0
    - recompile coq-mathcomp-fingroup  2.0.0*        [uses coq-mathcomp-ssreflect]
    - recompile coq-mathcomp-algebra   2.0.0*        [uses coq-mathcomp-fingroup]
    - recompile coq-mathcomp-solvable  2.0.0*        [uses coq-mathcomp-algebra]
    - recompile coq-mathcomp-field     2.0.0*        [uses coq-mathcomp-solvable]
    - recompile coq-mathcomp-character 2.0.0*        [uses coq-mathcomp-field]
  ===== 1 to install | 16 to recompile =====
```

In particular, see the `[upstream or system changes]`and the `recompile coq` patterns.

**Comment by JasonGross** [comment]
Could we have a job that runs on a schedule (say every night), does `opam update && opam install --dry-run coq | tee log` or something, and then greps the log for the pattern that indicates a rebuild is required and triggers one if necessary?

**Comment by erikmd** [comment]
@JasonGross
Yes, good point.

This issue (= follow-up of the Zulip discussion) was proposing a relatively-manual approach.
I agree it's possible to follow another approach, by automating a specific cron build.
In which case:
* it should rather been implemented in GitLab (e.g. with a `nightly` build in docker-keeper) because GitHub cron workflows automatically expire after 90 days.
* and the main question is: what would be the optimal script for this "detector"? given that dcker-coq currently contains 82 images and docker-mathcomp 71... and even if there's some approximation, we'd want to avoid false-negative.

**Comment by ejgallego** [comment]
Indeed this is a serious flaw of the opam package model...

On the other hand, these kind of changes should never happen, are a kind of last resort. Maybe the upcoming opam 2.2 forced more changes than usual?

I assume the system GMP didn't change.

Erik maybe we should open an issue on the opam repository as to involve the maintainers there.

For "upstream" changes the detection can be made pretty exact: check the checksum of the upstream opam file. This can automated using opam's API without having to rely on grep (please let's avoid parsing user messages)

**Comment by SkySkimmer** [comment]
What's the status of this?

**Comment by erikmd** [comment]
@SkySkimmer: meanwhile, @palmskog joined (who extensively use docker-mathcomp images).

But the issue shouldn't be closed yet IMHO.

BTW, the remarks by Emilio made me think that we could just as well automate these specific rebuilds.

1. For example, it is trivial to add a weekly (unconditional) rebuild of docker-rocq images.
   this is not done yet, but this wouldn't be so unusual, see e.g. `ocamlpro/ocaml` images that are weekly rebuilt.

2. Otherwise, it could be useful to add specific checks to reduce the CI load, while being more reactive.
   Emilio suggests we rely on opam's API to this aim, in order to be as robust as possible.

I won't have time to implement 2. before April, but let me know if ever you'd vote that I setup 1. now, until further notice.
Cc @Himito @Zimmi48 WDYT?

**Comment by Zimmi48** [comment]
> around several times per month, it happens some base opam packages are modified upstream, impyling coq is "rebuilt downstream" which thereby temporarily reduces the usefulness of the prebuilt coq. See the Example log below that shows the standard log pattern in this case.

Has this still been the case at this kind of frequency recently? If yes, then doing a weekly rebuild does seem quite justify.

If it's only once or twice a year, not so much.
